wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,336 @@
+   * @throws SQLException If an error occurs while attempting to establish a connection, or if we are connecting to a
+   *                      custom endpoint and we time out while waiting for custom endpoint info to be placed in the
+   *                      cache by the custom endpoint monitor.
+   */
+  protected Connection connectInternal(
+      HostSpec hostSpec,
+      Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    if (!this.rdsUtils.isRdsCustomClusterDns(hostSpec.getHost())) {
+      return connectFunc.call();
+    }
+
+    this.customEndpointHostSpec = hostSpec;
+    LOGGER.finest(
+        Messages.get(
+            "CustomEndpointPlugin.connectionRequestToCustomEndpoint", new Object[]{ hostSpec.getHost() }));
+
+    this.customEndpointId = this.rdsUtils.getRdsClusterId(customEndpointHostSpec.getHost());
+    if (StringUtils.isNullOrEmpty(customEndpointId)) {
+      throw new SQLException(
+          Messages.get(
+              "CustomEndpointPlugin.errorParsingEndpointIdentifier",
+              new Object[] {customEndpointHostSpec.getHost()}));
+    }
+
+    this.region = regionUtils.getRegion(this.customEndpointHostSpec.getHost(), props, REGION_PROPERTY.name);
+    if (this.region == null) {
+      throw new SQLException(
+          Messages.get(
+              "CustomEndpointPlugin.unableToDetermineRegion",
+              new Object[] {REGION_PROPERTY.name}));
+    }
+
+    CustomEndpointMonitor monitor = createMonitorIfAbsent(props);
+
+    // If needed, wait a short time for custom endpoint info to be discovered.
+    waitForCustomEndpointInfo(monitor);
+
+    return connectFunc.call();
+  }
+
+  /**
+   * Creates a monitor for the custom endpoint if it does not already exist.
+   *
+   * @param props The connection properties.
+   */
+  protected CustomEndpointMonitor createMonitorIfAbsent(Properties props) {
+    return monitors.computeIfAbsent(
+        this.customEndpointHostSpec.getHost(),
+        (customEndpoint) -> new CustomEndpointMonitorImpl(
+            this.pluginService,
+            this.customEndpointHostSpec,
+            this.customEndpointId,
+            this.region,
+            TimeUnit.MILLISECONDS.toNanos(CUSTOM_ENDPOINT_INFO_REFRESH_RATE_MS.getLong(props)),
+            this.rdsClientFunc
+        ),
+        TimeUnit.MILLISECONDS.toNanos(this.idleMonitorExpirationMs)
+    );
+  }
+
+
+
+  /**
+   * If custom endpoint info does not exist for the current custom endpoint, waits a short time for the info to be
+   * made available by the custom endpoint monitor. This is necessary so that other plugins can rely on accurate custom
+   * endpoint info. Since custom endpoint monitors and information are shared, we should not have to wait often.
+   */
+  protected void waitForCustomEndpointInfo(CustomEndpointMonitor monitor) throws SQLException {
+    boolean hasCustomEndpointInfo = monitor.hasCustomEndpointInfo();
+
+    if (!hasCustomEndpointInfo) {
```

Comment:
So we always wait if cache is empty? Is it so essential to wait for monitoring thread fetch custom endpoint definition? Can we make it configurable (wait/not wait)?

Reply:
The problem with not waiting is that, if the custom endpoint info isn't discovered yet, there will be no disallowed hosts. Which means that failover and the read write splitting plugin may connect to hosts outside of the custom endpoint. It seems like the user would not want this to happen if they are enabling this plugin. I believe the user will not have to wait long or often, on my machine it takes ~260ms for the API call to complete and the user will only need to wait when info isn't in the cache already (cache entries last 5 minutes but will be extended by active monitors). If you want I can make it configurable, but if the wait is disabled the custom endpoint plugin may allow connections to instances outside the endpoint.

Reply:
I'd prefer to delegate this decision to a user so a new configuration parameter would be great. Also I see that after calling `waitForCustomEndpointInfo()` the driver continues with `connectFunc` to get a connection. That means that custom endpoint info isn't directly required at this point.

Reply:
It might be required, there are two problem scenarios I can think of:
1. user has enableConnectFailover set to true, they try to connect but it fails and failover is kicked off, failover connects to a host outside of the custom endpoint info because we did not wait for the info to be found.
2. user connects successfully and then executes setReadOnly or hits failover while executing some other method, in both cases we could connect to an instance outside of the custom endpoint if we did not wait for the info to be found